tests/test_androidmodule_ctypes_finder.py

```diff
@@ -0,0 +1,158 @@
+
+# Import android._ctypes_library_finder module which we need to do isolated.
+# (since android.__init__ will crash if imported not on android!)
+import shutil
+import tempfile
+temp_d = tempfile.mkdtemp(prefix="p4a-test-ctypes-library-finder-")
+try:
+    import os
+    android_module_folder = os.path.abspath(os.path.join(
+        os.path.dirname(__file__),
+        "..", "pythonforandroid", "recipes", "android", "src"
+    ))
+
+    # Fake android module (with just android._ctypes_library_finder alone):
+    os.mkdir(os.path.join(temp_d, "android"))
+    with open(os.path.join(temp_d, "android", "__init__.py"), "w"):
+        pass
+    shutil.copyfile(os.path.join(android_module_folder,
+                                 "android/_ctypes_library_finder.py"),
+                    os.path.join(temp_d, "android",
+                                 "_ctypes_library_finder.py"))
+
+    # Import the module:
+    import sys
+    sys.path.insert(0, temp_d)
+    import android._ctypes_library_finder   # actual import
+    sys.path.remove(temp_d)
+finally:
+    shutil.rmtree(temp_d)
+
+
+# Fake jnius module with autoclass function:
+import sys
+import types
+module_name = 'jnius'
+bogus_module = types.ModuleType(module_name)
+sys.modules[module_name] = bogus_module
+bogus_module.autoclass = lambda name: None
+
+
+def _get_autoclass_with_fake_activity(returning_test_path="/testpath",
```

Comment:
Thanks for adding a docstring to give some explanation here. The first feeling is that we're over complicating things. I didn't deep dive into it yet, but I wasn't expecting such complexity to test what we're doing

Reply:
How else would you test `pyjnius` if not mock it to this extent? After all the android java api is otherwise not available

Reply:
Okay, so the `jnius` module itself is now placed a little nicer. But I'm not sure how to simplify specifically the `autoclass` mocking, I'm open for any suggestions!

Reply:
Yes I would mock it, but without implementing the methods. You just assert that the library was called with expected parameters because this is a unit test. So you assume the underlying library is already tested and behave consistently when called with some given params. So usually asserting on `call_args_list` is enough https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.call_args_list

Reply:
Hmmmmm but I did find a bug implementing the methods, pretty sure I wouldn't have found it without that. But as usual your call, but it seems to me it was a useful thing to do & have around, because otherwise truly testing this means firing up an android device which is a huge resistance to anyone ever doing it

Reply:
Yes I would definitely not have too much logic in the tests and kill that `autoclass_returning_graphical_activity` helper. You don't need it to actually go through the code path.
The idea would be to instead magic mock `jnius.autoclass.__init__` and makes it returns the activity class that contains one time the mActivity and the other time the mService. Then you verify that the activity was called with the correct parameters. So it should pretty much test the same as you did, but without building an helper that big

Reply:
I don't think I follow. How would you verify the correct parameters were called without implementing something similar? Wouldn't you need to also have some function that tracks what it was called with? Or does mock have something here I missed?

Also, at least the final call that returns the path needs to return a meaningful value, because otherwise the entire test for `test_find_library`

You'll probably have to give me some pointers for what mock functionality to google or a small example to make me understand how you want me to implement this, I'm afraid, because I'm not fully following right now :grimacing:

Reply:
My idea is to play with Mock.call_args_list to verify your mock got called with the expected parameters https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.call_args_list
I agree it's tricky to grasp at first and specially this case in not simple for different reasons, e.g. the lazy import, the top level module import and the jnius module not even being installed. So yes it's three things that makes it tricky to test it a straight forward way. And actually pyjnius could easily be a test dep. But we can also deal without it I guess.

Reply:
Regardig the test path being meaningful, yes for an integration test, not for a unit test.
Do we like integration test? Yes sometimes :) Do we like missing up with the file system in tests, probably not

Reply:
@AndreMiras it's a *temp folder*, it's not messing up anything calm down :joy: it's a useful and very simple test

I just tested call_args_list/mock_calls/method_calls and they work well enough, **but** how do you suggest I return that path in that final relevant inner nested call? I still see no way to avoid this nesting, so I'm not sure that helps with anything

Reply:
I wasn't kidding by the way when I said this caught a bug I didn't see before. unless you really want me to, I really would prefer not to cut even one of the test cases

Reply:
ohhhhh neat it can just be set nested and it stays there, that is coooool ok wait a sec, I can simplify this

Reply:
It's still messing with the file system even if it's a temp folder. I trust you could find a bug with this test, that's also the purpose. I'm not surprise you didn't find how to make the mock useful here because as I said it's not straightforward in our case for the reason I gave above. That's why I'll give it a try when I get a chance to. I can take a look tonight, but it's tight

Reply:
Okay I managed to simplify it a bit realizing I can just assign to some nested things right away. It's still not super short but way shorter than before :smile:

Reply:
Nice we're getting there, thanks.
Here's an example of what I'm thinking of when unit testing `get_activity_lib_dir()` for instance. Here I test only the first `return None` I'll write some more later:
```
def test_get_activity_lib_dir_autoclass_init_none():
    """
    Makes sure get_activity_lib_dir() returns `None` when `autoclass()`
    returns `None`.
    """
    m_jnius = mock.MagicMock()
    m_autoclass = mock.MagicMock()
    m_autoclass.return_value = None
    m_jnius.attach_mock(m_autoclass, 'autoclass')
    with mock.patch.dict('sys.modules', jnius=m_jnius):
        assert android._ctypes_library_finder.get_activity_lib_dir(
            'JavaClass'
        ) is None
        assert m_autoclass.call_args_list == [mock.call('JavaClass')]
```

Reply:
Wouldn't that make everything much longer though? When switching out `jnius` in every test rather than once at the top

Reply:
I like that argument check though, I added that one in now!

Reply:
Yes it's surely longer if you write a unit test per code path of your tested function. For instance if I want to write two more code path I can do:
```python
def test_get_activity_lib_dir_mactivity():
    """
    By default the mActivity is being accessed to retrieve the native library dir
    through the application context.
    """
    m_jnius = mock.MagicMock()
    m_autoclass = mock.MagicMock()
    m_jnius.attach_mock(m_autoclass, 'autoclass')
    with mock.patch.dict('sys.modules', jnius=m_jnius):
        m_app_context = m_autoclass().mActivity.getApplicationContext()
        m_app_info = m_app_context.getPackageManager().getApplicationInfo()
        m_app_info.nativeLibraryDir = '/path/to/lib'
        assert android._ctypes_library_finder.get_activity_lib_dir(
            'JavaClass'
        ) == '/path/to/lib'

def test_get_activity_lib_dir_mservice():
    """
    If the mActivity is `None` we fallback to the mService.
    """
    m_jnius = mock.MagicMock()
    m_autoclass = mock.MagicMock()
    m_autoclass().mActivity = None
    m_jnius.attach_mock(m_autoclass, 'autoclass')
    with mock.patch.dict('sys.modules', jnius=m_jnius):
        m_app_context = m_autoclass().mService.getApplicationContext()
        m_app_info = m_app_context.getPackageManager().getApplicationInfo()
        m_app_info.nativeLibraryDir = '/path/to/lib'
        assert android._ctypes_library_finder.get_activity_lib_dir(
            'JavaClass'
        ) == '/path/to/lib'
```
But of course you could also group these. But anyway it's not about the test being long, it's more about it not adding too much side logic that could make it unclear the code path we're testing

Reply:
Well the jnius module itself never changes in any way beyond the obvious autoclass assignment that is local to every test, so personally I'm not too sure that looks any better. But sure if you prefer the local `mock.patch.dict` I can change it

Reply:
So this is a draft to show the idea, but it could be much more comprehensive while still not having any `if` branching. See how in the second test `test_get_activity_lib_dir_mservice` we just make the `mActivity = None` to test another branching of our `get_activity_lib_dir_mservice()` code. Plus see then later in this same test we retrieve the app context through `mService` via `m_app_context = m_autoclass().mService.getApplicationContext()` in order to patch the `nativeLibraryDir` return value. This prove we went through the correct code path for this case also. I should have added comments even if it's a draft to make this `mService` case more explicit.
Also note in the same fashion we could have covered `if activity is None:` case of `get_activity_lib_dir()` if we wanted, which is not currently covered by your tests.
Anyway I agree said mock/testing would hardly be straightforward, but I would definitely be more comfortable if skip utils functions with code branching in our tests. Tests should always be a dumb as possible

Reply:
Well I usually prefer to patch as close as possible to where it's needed, just like I prefer to try/catch also as close as possible. But this wasn't the point of my test actually. My point is to show you can achieve what you were achieving without utils function containing if/ifelse/else which to me can make things look sketchy

Reply:
I revamped it a little according to your ideas, let me know what you think!

Reply:
Thanks!